### PR TITLE
Better expected sales calculation

### PR DESCRIPTION
**Pull request by spraot**
This solves #4

### CODE REVIEW

**Comment by spraot** [comment]
This only makes a difference for server add-ons.

**Comment by thecatontheflat** [comment]
Thank you @rachetfoot 

I will have a detailed look during this weekend

**Comment by thecatontheflat** [comment]
Expected amount of sales is different for my vendor (Cloud Add-Ons only). 

Before merge:
screenshot 2016-03-20 19 45 46

After:
screenshot 2016-03-20 19 45 52

Could you please explain the logic of "guessing"? In my current version it simply takes the last sale by a license (either monthly or annually) and projects it to the current month. For example, in this month there are 5 **expiring** licenses: 4 monthly subscriptions and 1 annual. The logic takes last sales of those licenses (4 monthly + 1 annual).

How does it work exactly in the proposed solution?
Do you think it should be possible to unify it for both server and cloud versions?

**Comment by spraot** [comment]
Well, I haven't properly handled the case of annual cloud licenses (I don't
have any), which I expect would explain the entire difference for you. I
take the current renewal prices for the add-ons rather than the last sale
price, but I couldn't see how to tell whether the last sale was annual or
monthly, so I assumed monthly. If there is no better solution, we could go
back to using the last sale price for cloud addons.

The proposed solution is unified for cloud and server versions, expect for
the annual/monthly issue I mentioned.

The reason for guessing is that only about 1% of my server users renew
exactly on the expiry date. I've fitted a normal distribution (bell curve)
to when people renew relative to license expiry date and base my estimate
on that. This is applied to all AUTO_QUOTE licenses, while I assume that
all AUTO_RENEW licenses will be renewed exactly on the expiry date (cloud
addons are always AUTO_RENEW). The normal distribution isn't a perfect fit,
but it's certainly a much better guess than the original.

**Comment by spraot** [comment]
I made a few updates and tested it for the past few months. The estimates looked fairly good. Let me know if it fixed the discrepancy you saw.

I also don't know whether the distribution parameters and renewal rate I selected are suitable for all vendors. They could potentially be a input parameter instead of constants.

**Comment by thecatontheflat** [comment]
@rachetfoot I am waiting until this issue is resolved:
https://ecosystem.atlassian.net/browse/AMKT-15693

Then I could test it properly. At this moment, I cannot rely on the AMKT data

I will merge it once I am able to get "clean" results. So far even my local version is different from what I have in production

**Comment by thecatontheflat** [comment]
I have this error when I try to load homepage:

screenshot 2016-04-02 11 40 39

**Comment by spraot** [comment]
I guess you have sales with a null maintenance end date. Is that expected?

File: src/AppBundle/Entity/LicenseRepository.php [hunk 1]
```diff
@@ -134,6 +134,91 @@ public function findRecent()
...
+            ->getQuery()
+            ->getResult();
+
+        $yesterday = new \DateTime('-1 days');
+        $endofmonth = new \DateTime('last day of this month');
+
+        $total = 0;
+        foreach ($licenses as $license) {
+            $lastSales = $this->getEntityManager()->getRepository('AppBundle:Sale')->findLastSalesByLicenses(array($license));
+            $lastSale = reset($lastSales);
```
Comment (thecatontheflat) [inline src/AppBundle/Entity/LicenseRepository.php:190]: When `$lastSales` is empty array, `reset()` returns false.

The login then fails later

Comment (spraot) [inline src/AppBundle/Entity/LicenseRepository.php:190]: I think I've fixed it now.

**Comment by thecatontheflat** [comment]
Hey @rachetfoot 
Sorry, I've been away for a while

I've checked the forecast today (first day of the month).
Current logic: `$6,564.75`
Logic in the PR: `$11,030.00`

I definitely like the latter :D But it seems that there is a problem somewhere

I have only cloud Add-Ons

File: src/AppBundle/Entity/LicenseRepository.php [hunk 2]
```diff
@@ -134,6 +134,95 @@ public function findRecent()
...
+
+            // Get user count
+            $userCnt = $license->getEdition();
+            $renewalLength = 12;
+            if ($userCnt == 'Subscription') {
+                if ($lastSale) {
+                    $userCnt = $lastSale['licenseSize'];
+                    if ($lastSale['maintenanceEndDate']->diff($lastSale['maintenanceStartDate'])->days < 45) {
+                        $renewalLength = 1;
+                    }
```
Comment (spraot) [inline src/AppBundle/Entity/LicenseRepository.php:206]: I'm guessing that the wrong estimate you see is due to reading the wrong renewalLength here. My php installation is messed up at the moment, so I can't look at this now, but this is where I think you should look if you have time.

**Comment by thecatontheflat** [comment]
Outdated

**Pull request closed without merging**